### ISSUE TITLE
Implement HAL semaphore using CUDA events

### ISSUE THREAD

**Issue by ThomasRaoux** [openxla/iree#4727]
IREE uses semaphores to synchronize host<->device and device<->device across command buffers. The only efficient way to synchronize different streams in cuda  is to use CUEvents. We need to find a way to emulate HAL semaphore using CUevent.

**Comment by ThomasRaoux** [comment]
FYI @benvanik, I'm still digging to figure this out but if you have already thought about, it would great to hear what you think.

**Comment by benvanik** [comment]
There's a few different ways of varying complexity :)

---

The most straightforward way is the stream memory operations: https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEMOP.html

These are a futex-like API that let you wait and write on memory addresses. Timeline semaphores *are* futexes, so this maps naturally - semaphore wait on value T is cuStreamWaitValue64(sema_ptr, T), and signal on value T+1 is cuStreamWriteValue64(sema_ptr, T). This is a ~20-line implementation: each sema gets a uint64_t slot in some table of addresses and you're done.

The gotcha here is that it says this is "disabled by default". I'm not sure what it means to enable it. Figuring out if we have support for it in the places we want to run today would be good, and even if it does require something to enable it we could use it to unblock the work and then come back and add another one of these options. This will always be the most efficient way of doing the semaphores and is perfect for our uses (if we can use it :).

---

Built on top of that functionality internally and not requiring enabling anything special is this:
https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EXTRES__INTEROP.html

This has the methods needed: cuSignalExternalSemaphoresAsync and cuWaitExternalSemaphoresAsync. They map 1:1 with Vulkan timeline semaphores.

Unfortunately I can't seem to figure out how to create one of those without using Vulkan 🤦 cuImportExternalSemaphore takes an fd/HANDLE of an existing timeline semaphore and I'm not sure you can just pass any old fd. The only example I see is using Vulkan interop to get them: https://sourcegraph.com/github.com/FFmpeg/FFmpeg/-/blob/libavutil/hwcontext_vulkan.c#L2422-2432

There's one lonely person here asking the same question: https://forums.developer.nvidia.com/t/how-to-create-a-semaphore-importable-by-cuimportexternalsemaphore/148212 (tried necroing the thread, will see if anyone notices it)

If we could figure that out (without requiring Vulkan) that'd be another ~20-line implementation.
It'd be worth finding someone from/friendly-with nvidia to see if there's a way.

---

We can simulate stream memory for some cases and have a heavier-weight approach for the others. Timeline semaphores have very strict semantics around signaling being monotonically increasing values (meaning that you only ever signal forward). It conversely has very loose requirements around waiting: waits are always strictly >= ("wait for the semaphore to be signaled to any value >= T"). This allows semaphores to be serialized such that a wait for T is allowed to block until T+123 - it's not great as the intent from the waiter was that they could begin doing something immediately at T+1 but had to wait until T+123 to go, but it's still correct. This means that a simple emulation may just pop off cuEventRecords for each signal with events from an event pool (or reuse the same event if there were no wait requests between the previous signal and the current one; cuEventRecord has those semantics). Waiters then would capture the most recent signal event if it was to a T >= the wait value and cuStreamWaitEvent on it. If the previous signal event was < T then a new event is allocated and is stashed to be inserted when a signal to any T >= is done.

This should be fairly small (~200 lines, just as an order of magnitude guess) and quick to get working at the cost of these out-of-order non-overlapping signal/waits possibly hurting concurrency. _But_ we don't currently schedule work like that from the compiler (and likely won't for many months) so I suspect an implementation like this would be good enough until then and still be correct (just not optimal) when we do start scheduling it. The upside is that semaphores are coarse-grained synchronization: ideally we submit quite a bit of work between each one, and overlap work where possible such that even if there is a latency involved there's still enough work pending to keep the device happy.

---

A bigger hammer is that we could implement them similar to how we do the fallback in Vulkan but we have more flexibility here because CUDA can do crazier things. Each timepoint in the timeline gets a CUevent from a pool to be waited/signaled, and waits/signals track all of those and sequence submissions to the stream based on them (the whole "serializing command queue" in the Vulkan HAL). This is the most code but is possible to implement with nothing special from the API. This approach has some major performance caveats that we didn't care to address on the Vulkan side because it's a fallback for deprecated stuff - we'd want to do something better here (where better == more in-depth use of the CUDA API). This introduces big bubbles in the device pipeline as host round-trips are required for all signals/waits and work submission has additional latency. I'm really hoping we could get one of the above ways working first :)

There's a few other approaches here involving various tricks of the API (emulating the stream memory operations by enqueuing kernels that just write a 64-bit value to memory, etc) to avoid the bubbles, but if we can get any of the above working first and consider this style a fallback for the fallback we could punt this until (or if) we ever need it.

---

Finally, we could change the HAL API. I'm not convinced it's worth it as the timeline semaphores are extremely useful simplifying assumptions for higher levels of the stack. The tradeoff here was that each HAL backend may have to do some backend-specific emulation of them but that how best to emulate them will dramatically differ across the backends and by keeping the magic in there we have the best hope of creating something efficient everywhere. It's also much easier to ensure program correctness with timeline semaphores and catch mistakes, which is useful when we have both user apps and compilers touching them.

We can have multiple code paths of the above, too. In the Vulkan backend this is like how we have a native timeline semaphore and an emulated one. Here in CUDA land we could have the memory_semaphore and external_semaphore and emulated_semaphore etc etc for example and switch which one we create at runtime based on what's available on the system.

**Comment by alliemartens** [comment]
@ThomasRaoux Stale P1 item here, please review when you get a chance!

**Comment by antiagainst** [comment]
Looking into implementing this for #13245; capturing current thoughts thus far:

[Stream memory operations ](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEMOP.html) match nicely here but we cannot use them, or use them as the only implementing mechanism, given the explicit warning there

> Improper use of these APIs may deadlock the application. Synchronization ordering established through these APIs is not visible to CUDA. CUDA tasks that are (even indirectly) ordered by these APIs should also have that order expressed with CUDA-visible dependencies such as events. This ensures that the scheduler does not serialize them in an improper order.

We'd need to interact with CUDA libraries and such so we'd need to be able to support `CUevent` anyway. So implementing IREE HAL semaphore with `CUevent` is a more suitable start--it gives us a widely applicable default; we can see how to optimize certain cases with other mechanisms later.

@benvanik I don't fully understand your "lightweight hammer" approach in the above. Would you mind to elaborate a bit? My current thoughts is go the "heavyweight hammer" approach like I'd previously for emulating timeline semaphore in Vulkan https://github.com/openxla/iree/pull/2208. It's much simpler though, given the semantics of `CUevent`.

`CUevent` is like Vulkan binary semaphore (if we ignore the elapsed time part, which we don't care here) but more flexible:
* It supports signal via device (`cuGraphAddEventRecordNode`), wait via host (`cuEventSynchronize`) or device (`cuGraphAddEventWaitNode`), which binary semaphore only supports device signal/wait. 
* `CUevent` can be signaled multiple times (although the captured workload may change) and waited multiple times. Binary semaphore is binary in signal and wait.
* Though we cannot perform wait before signal for `CUevent` too, given the spec "Before the first call to cuEventRecord(), an event represents an empty set of work, so for example cuEventQuery() would return CUDA_SUCCESS." This is the same for binary semaphores.

So the idea is basically have a timeline object with `CUevent` associated to different signaled timepoints on it. For timepoint wait we just grab whatever `CUevent` with a signaled value >= the waited value. There is no need to also use `VkFence` to track the progress of time points on different timelines like https://github.com/openxla/iree/pull/2208, given that `CUevent` can be directly handled on host. There is no need to be clever to avoid one waiter per signaled timepoint too. Though to respect the signal-before-wait behavior, we need to defer submissions in queue submit to inspect the wait semaphores first.

@benvanik please let me know whether this matches your thoughts. Also I recall you mentioned you have some existing utilities and some stuff in a branch? Would be great if I can get pointers and reuse some there! I found [`hal/utils/semaphore_base.h`](https://github.com/openxla/iree/blob/main/runtime/src/iree/hal/utils/semaphore_base.h) provides timeline and timepoint management. And [`local_task/task_semaphore.c`](https://github.com/openxla/iree/blob/main/runtime/src/iree/hal/drivers/local_task/task_semaphore.c) uses similar approach so probably worth to follow? Its associated [`task_queue.c`](https://github.com/openxla/iree/blob/main/runtime/src/iree/hal/drivers/local_task/task_queue.c) is a bit complicated yo go through though.

**Comment by antiagainst** [comment]
Thinking over this and here is the latest thoughts I have, with more concrete details for impl:

## Problem

Broadly as context, we have CPU and GPU operating async. For CPU, we can have multiple threads operating async, for GPU we can have multiple streams operating async. The communication and synchronization can happen between different CPU threads, between different GPU streams, and between a CPU thread and a GPU stream. IREE HAL semaphore provides a unified synchronization mechanism for all of these.

We are trying to implement IREE HAL semaphore mainly with `CUevent` from CUDA API. HAL semaphore is very flexible---it supports host/device signal/wait. While `CUevent` only supports device signal/wait and host wait. We need to design associated data structures and mechanisms to emulate the missing host signal and its interaction with other directions.

## Analysis

For a given HAL semaphore, looking at the four directions and see what's needed:

### CPU signal

A CPU  thread signals the semaphore timeline advancing to a new value. If there are

*  CPU waits: this is purely CPU side logic. We just need to use common CPU notification mechanisms. In IREE we have `iree_event_t` wrapping various low-level OS primitives for it. So we can just use that to represent a wait time point. Keep track of all CPU waited time points in the timeline. After a CPU signaling, go through the timeline and notify all earlier CPU waited values.
* GPU waits: given that there are no way we can signal `CUevent` on CPU, one way to go is to cache and defer the waiting submission batches by ourselves until CPU signals past the waited value. So this calls for a deferred/pending submission queue. 

### GPU signal

GPU signals can only be through `CUevent`, which has a binary state. We need to advance the timeline too. One way is to use `cuLaunchHostFunc` to perform the advance from the CPU side. This additionally would mean we can reuse the logic form CPU signal to unblock CPU waits.

For GPU waits, we can also leverage the same logic for CPU signaling and deferred GPU submitting dispatches. Though this is not nice for performance, given we don't want to involve CPU for GPU internal sync. We want to use `CUevent` for that:

*  So we can keep track of all GPU signaled values in the timeline. Once we see a GPU wait, try to scan the timeline to find a corresponding GPU signal advancing the timeline to pass the desired value, and use that for waiting instead. 
* We may not see GPU signal before seeing GPU wait requests, then we can also keep track of all GPU waits in the timeline too. Then later once see either CPU signal or GPU signal advancing past the waited value, we can handle them accordingly--submitting immediately or associating the `CUevent`. This would also guarantee the requirement of `CUevent`--recording should happen before waiting.
* We can use the same `CUevent` to unblock multiple GPU waits; that's allowed. Though it would mean we need to be careful regarding `CUevent` lifetime management. Here we can use some refcounting to see how many timepoints are using it. And automatically return to some pool once done. Having a pool of `CUevent` for recycling is good anyway.

Another problem is that per the `cuLaunchHostFunc` doc, "The function will be called after currently enqueued work and will block work added after it." We don't want the blocking behavior involving host. So we can use a dedicated `CUstream` for launching the host function, waiting on the `CUevent` from the original stream too. We can also handle resource deallocation together there.
  
## Data structures

So to summarize we pretty much need the following data structures:

* `iree_hal_cuda_event_t`: contains `CUevent` and an atomic counter. Used by device signal and wait timepoints. Associates with a pool--returns to the pool directly on refcounting == 0.
* `iree_hal_cuda_event_pool_t`: a list of the above objects. Avoid always creating and destroying CUDA objects.
* `iree_hal_cuda_host_wait_timepoint_t`: contains a `iree_event_t` for CPU wait. 
* `iree_hal_cuda_device_wait_timepoint_t`: contains an _optional_ `iree_hal_cuda_event_t`. If == nullptr then we wait on host, otherwise we can wait on the specified `CUevent`.
* `iree_hal_cuda_device_signal_timepoint_t`: conains an `iree_hal_cuda_event_t`. So that when later we see a GPU wait request we can pick this `CUevent` for waiting. 
* `iree_hal_cuda_timeline_semaphore_t`: contains a list of timepoints (can be any of the above three), a `CUstream` for host func launch and resource cleanup.
* `iree_hal_cuda_pending_submission_queue_t`: tracks all pending submissions with associated resources. Provides API to scan them and check dependency and perform submission accordingly.

## Import and export

For importing `CUevent`, we just need to insert a device wait timepoint warping the `CUevent`. For exporting, we just need to insert a device signal timepoint and exposing the underlying `CUevent`. It would enable interaction with external systems this way. Callers need to manually manage lifetime and such though.
